### PR TITLE
Only query for users identified by strings

### PR DESCRIPTION
**Pull request by indirect**

### CODE REVIEW

File: app/controllers/sessions_controller.rb [hunk 1]
```diff
@@ -19,6 +16,13 @@ def create
 
   private
 
+  def find_user(session)
+    who, pass = session.fetch(:who), session.fetch(:password)
+    if who.is_a?(String) && pass.is_a?(String)
```
Comment (bf4) [inline app/controllers/sessions_controller.rb]: 👍 Just noting here that `session.permit(:who)` would ensure `who` is in a list of [permitted scalars](https://github.com/rails/rails/blob/4-2-stable/actionpack/lib/action_controller/metal/strong_parameters.rb#L510-L531), such as String, but there's no way to restrict the list to just `String`, so we go type-checking route ourselves.

File: app/controllers/sessions_controller.rb [hunk 2]
```diff
@@ -19,6 +16,13 @@ def create
 
   private
 
+  def find_user(session)
+    who, pass = session.fetch(:who), session.fetch(:password)
```
Comment (bf4) [inline app/controllers/sessions_controller.rb]: Question: if either of these raise a `KeyError`, I assume there's some generic 500 response shown in the exceptions_app?

**Comment by bf4** [comment]
❗ Identical code is in the [v2/sessions_controller](https://github.com/rubygems/rubygems.org/blob/e00084c6a950ec44b07d80270c2a087e98ad71d8/app/controllers/api/v2/sessions_controller.rb#L2-L4)

**Comment by bf4** [comment]
Too bad I didn't finish https://github.com/rubygems/rubygems.org/pull/1202 :(

**Comment by qrush** [comment]
What's the motivation/error here? Seems like we should be able to use "standard" `params.require(:session).permit(:who, :password)` and rely on AC::Parameters to do the right thing.

**Comment by bf4** [comment]
@qrush I think https://github.com/rubygems/rubygems.org/pull/1255#discussion-diff-60634321

> 👍 Just noting here that `session.permit(:who)` would ensure `who` is in a list of [permitted scalars](https://github.com/rails/rails/blob/4-2-stable/actionpack/lib/action_controller/metal/strong_parameters.rb#L510-L531), such as String, but there's no way to restrict the list to just `String`, so we go type-checking route ourselves.

**Comment by segiddins** [comment]
What bug is this fixing? Do we need to add test coverage?

**Comment by bf4** [comment]
There was an exception reported: cannot call downcase on [1] from User.authenticate

B mobile phone

**Comment by indirect** [comment]
@segiddins I added a test

@qrush permit/require allows lots of [object types other than string](http://api.rubyonrails.org/classes/ActionController/Parameters.html), and based on the exception we saw where someone was trying to log in by sending a user that got parsed as an Array, I feel much more comfortable with completely avoiding the SQL query unless we are sure we got a string.

**Comment by indirect** [comment]
@bf4 thanks for pointing out the code duplication! that file... was creating the wrong constant, and also didn't get used at all. so I deleted it.

**Comment by indirect** [comment]
@dwradcliffe r?

**Comment by dwradcliffe** [comment]
:+1:

File: app/controllers/sessions_controller.rb [hunk 3]
```diff
@@ -19,6 +16,13 @@ def create
 
   private
 
+  def find_user(session)
+    who = session[:who].is_a?(String) && session.fetch(:who)
```
Comment (arthurnn) [inline app/controllers/sessions_controller.rb]: only if Strong parameters were actually strong!

**Comment by arthurnn** [comment]
🚢

**Pull request merged**